Cannot build quick perf from sources

I've cloned the repo and tried to build, but the build is failing with:
```
[INFO] --- jar:3.2.0:jar (default-jar) @ quick-perf-jvm-annotations ---
[INFO] Building jar: /home/qk/dev/demos/quickperf/jvm/jvm-annotations/target/quick-perf-jvm-annotations-1.1.0.jar
[INFO]
[INFO] --------------< org.quickperf:quick-perf-jfr-annotations >--------------
[INFO] Building quick-perf-jfr-annotations 1.1.0                         [6/40]
[INFO]   from jvm/jfr-annotations/pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] Artifact org.openjdk.jmc:flightrecorder.rules.jdk:pom:7.1.1 is present in the local repository, but cached from a remote repository ID that is unavailable in current build context, verifying that is downloadable from [Central (https://repo.maven.apache.org/maven2, default, releases), central (https://repo.maven.apache.org/maven2, default, releases)]
[INFO] Artifact org.openjdk.jmc:flightrecorder.rules.jdk:pom:7.1.1 is present in the local repository, but cached from a remote repository ID that is unavailable in current build context, verifying that is downloadable from [Central (https://repo.maven.apache.org/maven2, default, releases), central (https://repo.maven.apache.org/maven2, default, releases)]
[WARNING] The POM for org.openjdk.jmc:flightrecorder.rules.jdk:jar:7.1.1 is missing, no dependency information available
[INFO] Artifact org.openjdk.jmc:flightrecorder.rules.jdk:jar:7.1.1 is present in the local repository, but cached from a remote repository ID that is unavailable in current build context, verifying that is downloadable from [Central (https://repo.maven.apache.org/maven2, default, releases), central (https://repo.maven.apache.org/maven2, default, releases)]
[INFO] Artifact org.openjdk.jmc:flightrecorder.rules.jdk:jar:7.1.1 is present in the local repository, but cached from a remote repository ID that is unavailable in current build context, verifying that is downloadable from [Central (https://repo.maven.apache.org/maven2, default, releases), central (https://repo.maven.apache.org/maven2, default, releases)]
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for quick-perf 1.1.0:
[INFO]
[INFO] quick-perf ......................................... SUCCESS [  0.628 s]
[INFO] quick-perf-core .................................... SUCCESS [  3.174 s]
[INFO] quick-perf-jvm-parent .............................. SUCCESS [  0.010 s]
[INFO] quick-perf-jvm-core ................................ SUCCESS [  0.136 s]
[INFO] quick-perf-jvm-annotations ......................... SUCCESS [  1.542 s]
[INFO] quick-perf-jfr-annotations ......................... FAILURE [  0.025 s]
[INFO] quick-perf-sql-parent .............................. SKIPPED
[INFO] quick-perf-sql-annotations ......................... SKIPPED
[INFO] quick-perf-sql-memory-test-util .................... SKIPPED
[INFO] quick-perf-sql-hibernate-test-util ................. SKIPPED
[INFO] quick-perf-junit4-parent ........................... SKIPPED
[INFO] quick-perf-junit4 .................................. SKIPPED
[INFO] quick-perf-sql-memory-test ......................... SKIPPED
[INFO] quick-perf-junit4-5-jvm-test ....................... SKIPPED
[INFO] quick-perf-junit4-12-jvm-test ...................... SKIPPED
[INFO] quick-perf-junit4-13-jvm-test ...................... SKIPPED
[INFO] quick-perf-junit4-13-1-jvm-test .................... SKIPPED
[INFO] quick-perf-junit4-13-2-jvm-test .................... SKIPPED
[INFO] quick-perf-junit4-sql-test ......................... SKIPPED
[INFO] quick-perf-junit5-parent ........................... SKIPPED
[INFO] quick-perf-junit5-test-util ........................ SKIPPED
[INFO] quick-perf-junit5 .................................. SKIPPED
[INFO] quick-perf-junit5-jvm-test ......................... SKIPPED
[INFO] quick-perf-junit5-sql-test ......................... SKIPPED
[INFO] quick-perf-testng-parent ........................... SKIPPED
[INFO] quick-perf-testng-test-util ........................ SKIPPED
[INFO] quick-perf-testng .................................. SKIPPED
[INFO] quick-perf-testng-sql-test ......................... SKIPPED
[INFO] quick-perf-testng-jvm-test ......................... SKIPPED
[INFO] quick-perf-spring .................................. SKIPPED
[INFO] quick-perf-sql-spring4 ............................. SKIPPED
[INFO] quick-perf-sql-spring5 ............................. SKIPPED
[INFO] quick-perf-junit4-spring-base-tests ................ SKIPPED
[INFO] quick-perf-junit4-spring3 .......................... SKIPPED
[INFO] quick-perf-junit4-spring4 .......................... SKIPPED
[INFO] quick-perf-junit4-spring5 .......................... SKIPPED
[INFO] quick-perf-springboot1-sql-starter ................. SKIPPED
[INFO] quick-perf-springboot2-sql-starter ................. SKIPPED
[INFO] quick-perf-junit4-spring-boot-test ................. SKIPPED
[INFO] quick-perf-bom ..................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  5.909 s
[INFO] Finished at: 2024-03-16T21:37:44+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project quick-perf-jfr-annotations: Could not resolve dependencies for project org.quickperf:quick-perf-jfr-annotations:jar:1.1.0: The following artifacts could not be resolved: org.openjdk.jmc:flightrecorder.rules.jdk:jar:7.1.1 (present, but unavailable): org.openjdk.jmc:flightrecorder.rules.jdk:jar:7.1.1 was not found in https://repo.maven.apache.org/maven2 during a previous attempt. This failure was cached in the local repository and resolution is not reattempted until the update interval of Central has elapsed or updates are forced -> [Help 1]

```

and after removing the jmc artifact from the local repo:

```
[ERROR] Failed to execute goal on project quick-perf-jfr-annotations: Could not resolve dependencies for project org.quickperf:quick-perf-jfr-annotations:jar:1.1.0: The following artifacts could not be resolved: org.openjdk.jmc:flightrecorder.rules.jdk:jar:7.1.1 (absent): Could not transfer artifact org.openjdk.jmc:flightrecorder.rules.jdk:jar:7.1.1 from/to AdoptOpenJDK (https://adoptopenjdk.jfrog.io/adoptopenjdk/jmc-libs): status code: 409, reason phrase: Conflict (409) -> [Help 1]

```

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -8,18 +8,19 @@ jobs:
     name: Java ${{ matrix.java }}
     strategy:
       matrix:
-        java: [8, 11, 12, 13, 14, 15, 16, 17]
+        java: [8, 11, 17]
     env:
       REPO_SLUG: ${{ github.repository }}
       BRANCH: ${{ github.head_ref }}
     steps:
-    - uses: actions/checkout@v2
+    - uses: actions/checkout@v4
     - name: Set up JDK ${{ matrix.java }}
-      uses: actions/setup-java@v1
+      uses: actions/setup-java@v4
       with:
         java-version: ${{ matrix.java }}
+        distribution: temurin
     - name: Cache Maven dependencies
-      uses: actions/cache@v2
+      uses: actions/cache@v4
       with:
         path: ~/.m2
         key: ${{ runner.os }}-m2-${{ hashFiles('**/pom.xml') }}
@@ -37,13 +38,14 @@ jobs:
       REPO_SLUG: ${{ github.repository }}
       BRANCH: ${{ github.head_ref }}
     steps:
-    - uses: actions/checkout@v2
+    - uses: actions/checkout@v4
     - name: Set up JDK ${{ matrix.java }}
-      uses: actions/setup-java@v1
+      uses: actions/setup-java@v4
       with:
         java-version: ${{ matrix.java }}
+        distribution: temurin
     - name: Cache Maven dependencies
-      uses: actions/cache@v2
+      uses: actions/cache@v4
       with:
         path: ~/.m2
         key: ${{ runner.os }}-m2-${{ hashFiles('**/pom.xml') }}
@@ -61,13 +63,14 @@ jobs:
       REPO_SLUG: ${{ github.repository }}
       BRANCH: ${{ github.head_ref }}
     steps:
-    - uses: actions/checkout@v2
+    - uses: actions/checkout@v4
     - name: Set up JDK ${{ matrix.java }}
-      uses: actions/setup-java@v1
+      uses: actions/setup-java@v4
       with:
         java-version: ${{ matrix.java }}
+        distribution: temurin
     - name: Cache Maven dependencies
-      uses: actions/cache@v2
+      uses: actions/cache@v4
       with:
         path: ~/.m2
         key: ${{ runner.os }}-m2-${{ hashFiles('**/pom.xml') }}
@@ -85,13 +88,14 @@ jobs:
       REPO_SLUG: ${{ github.repository }}
       BRANCH: ${{ github.head_ref }}
     steps:
-      - uses: actions/checkout@v2
+      - uses: actions/checkout@v4
       - name: Set up JDK ${{ matrix.java }}
-        uses: actions/setup-java@v1
+        uses: actions/setup-java@v4
         with:
           java-version: ${{ matrix.java }}
+          distribution: temurin
       - name: Cache Maven dependencies
-        uses: actions/cache@v2
+        uses: actions/cache@v4
         with:
           path: ~/.m2
           key: ${{ runner.os }}-m2-${{ hashFiles('**/pom.xml') }}

--- a/junit4/junit4-12-jvm-test/pom.xml
+++ b/junit4/junit4-12-jvm-test/pom.xml
@@ -21,9 +21,9 @@
     <artifactId>quick-perf-junit4-12-jvm-test</artifactId>
 
     <properties>
-        <maven.compiler.source>1.7</maven.compiler.source>
-        <maven.compiler.target>1.7</maven.compiler.target>
-        <dependencies.max.jdk.version>1.7</dependencies.max.jdk.version>
+        <maven.compiler.source>1.8</maven.compiler.source>
+        <maven.compiler.target>1.8</maven.compiler.target>
+        <dependencies.max.jdk.version>1.8</dependencies.max.jdk.version>
         <jar.skipIfEmpty>true</jar.skipIfEmpty>
         <gpg.skip>true</gpg.skip>
         <skipNexusStagingDeployMojo>true</skipNexusStagingDeployMojo>

--- a/junit5/junit5-jvm-test/pom.xml
+++ b/junit5/junit5-jvm-test/pom.xml
@@ -21,6 +21,9 @@
     <artifactId>quick-perf-junit5-jvm-test</artifactId>
 
     <properties>
+        <maven.compiler.source>1.8</maven.compiler.source>
+        <maven.compiler.target>1.8</maven.compiler.target>
+        <dependencies.max.jdk.version>1.8</dependencies.max.jdk.version>
         <jar.skipIfEmpty>true</jar.skipIfEmpty>
         <maven.install.skip>true</maven.install.skip>
         <maven.deploy.skip>true</maven.deploy.skip>

--- a/junit5/junit5-test-util/pom.xml
+++ b/junit5/junit5-test-util/pom.xml
@@ -24,6 +24,7 @@
         <maven.deploy.skip>true</maven.deploy.skip>
         <gpg.skip>true</gpg.skip>
         <skipNexusStagingDeployMojo>true</skipNexusStagingDeployMojo>
+        <allowIncompleteProjects>true</allowIncompleteProjects>
     </properties>
 
     <build>

--- a/jvm/jfr-annotations/pom.xml
+++ b/jvm/jfr-annotations/pom.xml
@@ -22,6 +22,9 @@
     <artifactId>quick-perf-jfr-annotations</artifactId>
 
     <properties>
+        <maven.compiler.source>1.8</maven.compiler.source>
+        <maven.compiler.target>1.8</maven.compiler.target>
+        <dependencies.max.jdk.version>1.8</dependencies.max.jdk.version>
         <javadoc-subpackages>org.quickperf.jvm.jfr.annotation</javadoc-subpackages>
     </properties>
 
@@ -39,7 +42,7 @@
         <dependency>
             <groupId>org.openjdk.jmc</groupId>
             <artifactId>flightrecorder.rules.jdk</artifactId>
-            <version>7.1.1</version>
+            <version>8.3.1</version>
         </dependency>
         <dependency>
             <groupId>org.jsoup</groupId>

--- a/jvm/jfr-annotations/src/main/java/org/quickperf/jvm/jmcrule/JmcRuleCountMeasureExtractor.java
+++ b/jvm/jfr-annotations/src/main/java/org/quickperf/jvm/jmcrule/JmcRuleCountMeasureExtractor.java
@@ -35,6 +35,7 @@ import java.util.stream.Collectors;
 public class JmcRuleCountMeasureExtractor implements ExtractablePerformanceMeasure<JfrRecording, JmcRulesMeasure> {
 
     public static final JmcRuleCountMeasureExtractor INSTANCE = new JmcRuleCountMeasureExtractor();
+    //We need to ignore some rules which appear to throw NullPointerException in org.openjdk.jmc:flightrecorder.rules.jdk:8.x.y
     private static final List<Class<? extends IRule>> IGNORED_RULES = Arrays.asList(
             HighJvmCpuRule.class,
             GcStallRule.class,

--- a/jvm/jfr-annotations/src/main/java/org/quickperf/jvm/jmcrule/JmcRuleCountMeasureExtractor.java
+++ b/jvm/jfr-annotations/src/main/java/org/quickperf/jvm/jmcrule/JmcRuleCountMeasureExtractor.java
@@ -13,33 +13,44 @@
 package org.quickperf.jvm.jmcrule;
 
 import org.openjdk.jmc.common.item.IItemCollection;
+import org.openjdk.jmc.common.unit.IQuantity;
 import org.openjdk.jmc.common.util.IPreferenceValueProvider;
-import org.openjdk.jmc.flightrecorder.rules.IRule;
-import org.openjdk.jmc.flightrecorder.rules.Result;
-import org.openjdk.jmc.flightrecorder.rules.RuleRegistry;
-import org.openjdk.jmc.flightrecorder.rules.Severity;
+import org.openjdk.jmc.flightrecorder.rules.*;
+import org.openjdk.jmc.flightrecorder.rules.jdk.cpu.HighJvmCpuRule;
+import org.openjdk.jmc.flightrecorder.rules.jdk.memory.GarbageCollectionInfoRule;
+import org.openjdk.jmc.flightrecorder.rules.jdk.memory.GcStallRule;
+import org.openjdk.jmc.flightrecorder.rules.jdk.memory.HeapInspectionRule;
+import org.openjdk.jmc.flightrecorder.rules.jdk.memory.SystemGcRule;
 import org.quickperf.ExtractablePerformanceMeasure;
 import org.quickperf.jvm.jfr.JfrRecording;
 import org.quickperf.unit.Count;
 
 import java.io.PrintWriter;
 import java.io.StringWriter;
-import java.util.ArrayList;
-import java.util.List;
+import java.util.*;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.RunnableFuture;
+import java.util.stream.Collectors;
 
 public class JmcRuleCountMeasureExtractor implements ExtractablePerformanceMeasure<JfrRecording, JmcRulesMeasure> {
 
     public static final JmcRuleCountMeasureExtractor INSTANCE = new JmcRuleCountMeasureExtractor();
+    private static final List<Class<? extends IRule>> IGNORED_RULES = Arrays.asList(
+            HighJvmCpuRule.class,
+            GcStallRule.class,
+            GarbageCollectionInfoRule.class,
+            HeapInspectionRule.class,
+            SystemGcRule.class
+    );
 
-    private JmcRuleCountMeasureExtractor() {}
+    private JmcRuleCountMeasureExtractor() {
+    }
 
     @Override
     public JmcRulesMeasure extractPerfMeasureFrom(JfrRecording jfrRecording) {
 
         IItemCollection jfrEvents = jfrRecording.getJfrEvents();
-        List<Result> ruleEvaluations = evaluateJmcRules(jfrEvents);
+        List<IResult> ruleEvaluations = evaluateJmcRules(jfrEvents);
 
         List<Count> jmcRules = buildJmcRuleCountsFrom(ruleEvaluations);
 
@@ -47,12 +58,16 @@ public class JmcRuleCountMeasureExtractor implements ExtractablePerformanceMeasu
 
     }
 
-    private List<Result> evaluateJmcRules(IItemCollection jfrEvents) {
-        List<Result> ruleEvaluations = new ArrayList<>();
-        for (IRule rule : RuleRegistry.getRules()) {
-            RunnableFuture<Result> future = rule.evaluate(jfrEvents, IPreferenceValueProvider.DEFAULT_VALUES);
+    private List<IResult> evaluateJmcRules(IItemCollection jfrEvents) {
+        List<IResult> ruleEvaluations = new ArrayList<>();
+        Collection<IRule> rules = RuleRegistry.getRules().stream()
+                .filter(r -> !IGNORED_RULES.contains(r.getClass()))
+                .collect(Collectors.toList());
+        for (IRule rule : rules) {
+            RunnableFuture<IResult> future = rule.createEvaluation(jfrEvents,
+                    IPreferenceValueProvider.DEFAULT_VALUES, new ResultProvider());
             future.run();
-            Result result;
+            IResult result;
             try {
                 result = future.get();
             } catch (InterruptedException | ExecutionException e) {
@@ -63,11 +78,11 @@ public class JmcRuleCountMeasureExtractor implements ExtractablePerformanceMeasu
         return ruleEvaluations;
     }
 
-    private List<Count> buildJmcRuleCountsFrom(List<Result> ruleEvaluations) {
+    private List<Count> buildJmcRuleCountsFrom(List<IResult> ruleEvaluations) {
         List<Count> jmcRules = new ArrayList<>();
-        for (Result ruleEvaluation : ruleEvaluations) {
+        for (IResult ruleEvaluation : ruleEvaluations) {
             Count ruleScore = buildJmcRuleCountFrom(ruleEvaluation);
-            if(!ruleToExclude(ruleScore)) {
+            if (!ruleToExclude(ruleScore)) {
                 jmcRules.add(ruleScore);
             }
         }
@@ -76,21 +91,25 @@ public class JmcRuleCountMeasureExtractor implements ExtractablePerformanceMeasu
 
     private boolean ruleToExclude(Count ruleScore) {
         String ruleDescription = ruleScore.getComment();
-        return     ruleDescription.contains("Rule: TLAB Allocation Ratio")
-                || ruleDescription.contains("Rule: Competing Processes")
-                || ruleDescription.contains("Rule: Command Line Options Check")
-                || ruleDescription.contains("Rule: Metaspace Live Set Trend");
+        return ruleDescription.contains("Rule: TLAB Allocation Ratio")
+               || ruleDescription.contains("Rule: Competing Processes")
+               || ruleDescription.contains("Rule: Command Line Options Check")
+               || ruleDescription.contains("Rule: Metaspace Live Set Trend");
     }
 
-    private Count buildJmcRuleCountFrom(Result result) {
+    private Count buildJmcRuleCountFrom(IResult result) {
         StringWriter stringWriter = new StringWriter();
 
         PrintWriter printWriter = new PrintWriter(stringWriter);
         printWriter.println("Rule: " + result.getRule().getName());
-        printWriter.println("Severity: " + Severity.get(result.getScore()));
-        long score = (long) result.getScore();
+        printWriter.println("Severity: " + result.getSeverity());
+        long score = Optional.ofNullable(result.getResult(TypedResult.SCORE))
+                .map(IQuantity::longValue)
+                .orElse(-1L);
         printWriter.println("Score: " + score);
-        String longDescriptionAsHtml = result.getLongDescription();
+        final String longDescriptionAsHtml = Optional.ofNullable(result.getSummary()).orElse("")
+                + Optional.ofNullable(result.getExplanation()).orElse("")
+                + Optional.ofNullable(result.getSolution()).orElse("");
         String textDesc = HtmlToPlainTextTransformer.INSTANCE.convertHtmlToPlainText(longDescriptionAsHtml);
         printWriter.println("Message: " + textDesc);
 

--- a/pom.xml
+++ b/pom.xml
@@ -49,6 +49,7 @@
         <commons-dbcp.version>1.4</commons-dbcp.version>
         <maven-jar-plugin.version>3.2.0</maven-jar-plugin.version>
         <maven-surefire-plugin.version>2.22.2</maven-surefire-plugin.version>
+        <maven-enforcer-plugin.version>3.5.0</maven-enforcer-plugin.version>
         <logback-classic.version>1.1.0</logback-classic.version>
         <jcl-over-slf4j.version>1.7.25</jcl-over-slf4j.version>
         <javadoc-subpackages />
@@ -112,8 +113,9 @@
                 </configuration>
             </plugin>
             <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-enforcer-plugin</artifactId>
-                <version>1.4.1</version>
+                <version>${maven-enforcer-plugin.version}</version>
                 <executions>
                     <execution>
                         <id>enforce-bytecode-version</id>
@@ -138,7 +140,7 @@
                     <dependency>
                         <groupId>org.codehaus.mojo</groupId>
                         <artifactId>extra-enforcer-rules</artifactId>
-                        <version>1.2</version>
+                        <version>1.9.0</version>
                     </dependency>
                 </dependencies>
             </plugin>
@@ -288,14 +290,6 @@
             <id>Central</id>
             <name>Central Repository</name>
             <url>https://repo.maven.apache.org/maven2</url>
-            <snapshots>
-                <enabled>false</enabled>
-            </snapshots>
-        </repository>
-        <repository>
-            <id>AdoptOpenJDK</id>
-            <name>AdoptOpenJDK</name>
-            <url>https://adoptopenjdk.jfrog.io/adoptopenjdk/jmc-libs</url>
             <snapshots>
                 <enabled>false</enabled>
             </snapshots>

--- a/spring/junit4-spring-base-tests/pom.xml
+++ b/spring/junit4-spring-base-tests/pom.xml
@@ -26,6 +26,7 @@
         <maven.deploy.skip>true</maven.deploy.skip>
         <gpg.skip>true</gpg.skip>
         <skipNexusStagingDeployMojo>true</skipNexusStagingDeployMojo>
+        <allowIncompleteProjects>true</allowIncompleteProjects>
     </properties>
 
     <dependencies>

--- a/sql/sql-hibernate-test-util/pom.xml
+++ b/sql/sql-hibernate-test-util/pom.xml
@@ -26,6 +26,7 @@
         <maven.deploy.skip>true</maven.deploy.skip>
         <gpg.skip>true</gpg.skip>
         <skipNexusStagingDeployMojo>true</skipNexusStagingDeployMojo>
+        <allowIncompleteProjects>true</allowIncompleteProjects>
     </properties>
 
     <dependencies>

--- a/sql/sql-memory-test-util/pom.xml
+++ b/sql/sql-memory-test-util/pom.xml
@@ -27,6 +27,7 @@
         <maven.deploy.skip>true</maven.deploy.skip>
         <gpg.skip>true</gpg.skip>
         <skipNexusStagingDeployMojo>true</skipNexusStagingDeployMojo>
+        <allowIncompleteProjects>true</allowIncompleteProjects>
     </properties>
 
     <dependencies>

--- a/testng/testng-jvm-test/pom.xml
+++ b/testng/testng-jvm-test/pom.xml
@@ -20,6 +20,9 @@
     <artifactId>quick-perf-testng-jvm-test</artifactId>
 
     <properties>
+        <maven.compiler.source>1.8</maven.compiler.source>
+        <maven.compiler.target>1.8</maven.compiler.target>
+        <dependencies.max.jdk.version>1.8</dependencies.max.jdk.version>
         <jar.skipIfEmpty>true</jar.skipIfEmpty>
         <gpg.skip>true</gpg.skip>
         <skipNexusStagingDeployMojo>true</skipNexusStagingDeployMojo>

--- a/testng/testng-test-util/pom.xml
+++ b/testng/testng-test-util/pom.xml
@@ -26,6 +26,7 @@
         <gpg.skip>true</gpg.skip>
         <skipNexusStagingDeployMojo>true</skipNexusStagingDeployMojo>
         <maven.deploy.skip>true</maven.deploy.skip>
+        <allowIncompleteProjects>true</allowIncompleteProjects>
     </properties>
 
     <dependencies>

